### ISSUE TITLE
Large Collections

### ISSUE THREAD

**Issue by Developer-Alexander** [dotnet/runtime#46586]
## Background and Motivation
Generic collection types like Arrays, Lists and Dictionary are limited to roughly 2.1 billion elements even on 64 bit machines.
That's a very annoying limitation when you want to work on very large sets of data.
On a 64 bit machine you should not need to handle your collections like you are on a 32 bit machine.

## Proposed API

Introduce new large, generic collection types and interfaces like `LargeArray<T>`, `LargeList<T>` and `LargeDictionary<TKey, TValue>` which can handle up to 2^64 elements.

``` C#
public interface ILargeList<T>
{
    ulong Count { get; }

    T this[ulong index] { get; set; }
}

public interface ILargeDictionary<TKey, TValue>
{
    ulong Count { get; }

    TValue this[TKey key] { get; set; }

    bool Remove(TKey key);
}

public class LargeArray<T>: ILargeList<T>, IEnumerable<T> {}

public class LargeList<T>: ILargeList<T>, IEnumerable<T> {}

public class LargeDictionary<TKey, TValue> : ILargeDictionary<TKey, TValue> {}

```

These definitions are representative for the entire set of generic collections and are not meant to be a complete definition.

To implement a Dictionary that can handle up to 2^64 elements it might be necessary to introduce a version of `object.GetHashCode()` which returns a 64 bit hash code.

## Usage Examples
The usage of large collections should be the same like todays collections beside being able to handle a bigger number of elements.

## Alternative Designs

Alternatively existing collection types and interfaces could be extended to support a bigger maximum number of elements.

## Risks
The introduction of large collections should be a compatible extension to the framework.

**Comment by huoyaoyuan** [comment]
There are already many discussions over large arrays and collections (effectively, `nint/nuint Length`.

You are free to implement large collections with long length, but such interface can hardly contribute to the ecosystem, because int length was assumed almost everywhere except streams.

> Alternatively existing collection types and interfaces could be extended to support a bigger maximum number of elements.

Huge collections implies another problem. Many collections uses continuous memory. It will become impossible for GC if being so large.

**Comment by eiriktsarpalis** [comment]
> Huge collections implies another problem. Many collections uses continuous memory. It will become impossible for GC if being so large.

Presumably though classes implementing that interface could use sharding?

**Comment by Developer-Alexander** [comment]
It is possible to implement LargeArrays or LargeLists with arrays of arrays. However I think this is something that should be provided by the framework.

But what you cant build on your own (at least for my knowledge) is a LargeDictionary since the framework does not provide an API for a 64 bit HashCode of objects. Last time I cheked HashCodes are calculated as 64 Bit values and then they are cropped to 32 bit before they are returned by `object.GetHashCode()`.

Would you consider to introduce a `object.GetHashCode64()` at least?

**Comment by huoyaoyuan** [comment]
> Would you consider to introduce a `object.GetHashCode64()` at least?

There's likely never chance to introduce new member in `object`.

**Comment by svick** [comment]
> > Would you consider to introduce a `object.GetHashCode64()` at least?
> 
> There's likely never chance to introduce new member in `object`.

Since the request seems to be about reference equality and you can use [`RuntimeHelpers.GetHashCode()`](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.gethashcode?WT.mc_id=DOP-MVP-5002735) for that, would it make sense to add `RuntimeHelpers.GetHashCode64()` instead?

**Comment by Clockwork-Muse** [comment]
@svick - 
> Since the request seems to be about reference equality and you can use `RuntimeHelpers.GetHashCode()` for that, would it make sense to add `RuntimeHelpers.GetHashCode64()` instead?

The documentation for `RuntimeHelpers.GetHashCode()` states that different hash codes can be returned for equal-value objects, which would break a dictionary.

@Developer-Alexander 
> But what you cant build on your own (at least for my knowledge) is a LargeDictionary since the framework does not provide an API for a 64 bit HashCode of objects. Last time I cheked HashCodes are calculated as 64 Bit values and then they are cropped to 32 bit before they are returned by `object.GetHashCode()`.

There's a couple of things you should consider here:
 1) Dictionaries "bucket" (group) items based on their hash code.  Unless you have _significantly_ more than 2 billion elements the extra bit width doesn't make a difference.
 2) You're forgetting about overrides of `GetHashCode()` - if the 64 bit version isn't based on the same information then things break in surprising ways.  

You _can_ build a large dictionary for yourself, especially if you know your key - you don't rely on the runtime-provided hash code at all.  This is actually more likely since if the collection is of any "real" size you likely need to be disposing objects/releasing memory after having persisted the object somehow - if the object is later deserialized the runtime hashcode will have changed.  Either you need to have your key objects all implement a specific interface with a method to calculate the 64-bit hash, or you have the dictionary take an equivalent to `IComparer` that will calculate the hash code and equality as necessary.

**Comment by svick** [comment]
@Clockwork-Muse Yes, I didn't mean that you could just replace `object.GetHashCode()` by `RuntimeHelpers.GetHashCode64()`. But you could create a consistent reference-equality-based 64-bit equality comparer from `object.ReferenceEquals()` and `RuntimeHelpers.GetHashCode64()`. So it could be a useful primitive for some large dictionaries. (And it's not something you can easily build yourself, unlike 64-bit value equality.)

**Comment by Developer-Alexander** [comment]
> 
> 
> > Would you consider to introduce a `object.GetHashCode64()` at least?
> 
> There's likely never chance to introduce new member in `object`.

And why?

For me it does not matter where the 64 bit Hash does come from. (If it comes from object or RuntimeHelpers...) But I really would appreciate to have at least one option.

**Comment by Clockwork-Muse** [comment]
Because introducing a new member to `object` has the potential to break everything, since it's the base type (compile would fail because of missing `override` if a class declares its own `GetHashCode64()`).  Besides any other concerns.

> For me it does not matter where the 64 bit Hash does come from. (If it comes from object or RuntimeHelpers...) 

You can use `RuntimeHelpers` now, but I actually disagree with @svick here - I personally believe that for dictionaries of that size value-based keys are **far** more likely/used (and thus `RuntimeHelpers.GetHashCode64()` could not be used), and the few times reference-based keys are used the collection size will be far smaller (and thus a 64-bit hash would not be necessary).

**Comment by danmoseley** [comment]
> for dictionaries of that size value-based keys are far more likely/used... the few times reference-based keys are used the collection size will be far smaller (and thus a 64-bit hash would not be necessary).

This is my assumption too. Plus, if your application is dealing with such huge maps, you are likely going to be optimizing heavily, and would likely want to invest in your own collection type optimized for your use case. If you simply care about size, you can easily multiplex several dictionaries by using one or more bits from the hash code. @Developer-Alexander are those fair assumptions in your use case?

**Comment by Developer-Alexander** [comment]
My use case was processing Open Street Map which is structured as a huge number of small objects. For performace reasons I needed a lot of fast lookups. (uint64 or uint32 id for example)

Dealing with this the size limitation of arrays became a problem and so did the limited capacity of dictionaries.

When you want a hash based dictionary and you hash is limited to 32 bit then you will end up in a linear search very quickly because many items will have the same hash. As far as I know there is no other way than using a bigger hash value to prevent linear search.

**Comment by danmoseley** [comment]
@Developer-Alexander  got it. Is a value-based hash feasible in your use case? As opposed to a reference one where it is fundamentally implemented by the runtime, is stored in the syncblock, effectively growing the object size, etc.
